### PR TITLE
[SR-1738] add_swift_library takes SHARED/STATIC arg

### PR DESCRIPTION
**Pull request by modocache**
#### What's in this pull request?

As a first step to allowing the build script to build _only_ static library versions of the stdlib (see #2431), change `add_swift_library` such that callers must pass in `SHARED`, `STATIC`, or `OBJECT_LIBRARY`.

Ideally, only these flags would be used to determine whether to build shared, static, or object libraries, but that is not currently the case -- `add_swift_library` also checks whether the library `IS_STDLIB` before performing certain additional actions. This will be cleaned up in a future commit.

/cc @gribozavr 
#### ~~Resolved~~ Related bug number: ([SR-1738](https://bugs.swift.org/browse/SR-1738))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by modocache** [comment]
@swift-ci please test

File: cmake/modules/AddSwift.cmake [hunk 1]
```diff
@@ -473,6 +477,11 @@ function(_add_swift_library_single target name)
   if("${SWIFTLIB_SINGLE_INSTALL_IN_COMPONENT}" STREQUAL "")
     message(FATAL_ERROR "INSTALL_IN_COMPONENT is required")
   endif()
+
+  if(NOT SWIFTLIB_SINGLE_SHARED AND NOT SWIFTLIB_SINGLE_STATIC AND NOT SWIFTLIB_SINGLE_OBJECT_LIBRARY)
```
Comment (gribozavr) [inline cmake/modules/AddSwift.cmake]: 80 columns, please.

**Comment by gribozavr** [comment]
LGTM, but I wanted to ask you what is your long-term plan about `${SWIFT_STDLIB_LIBRARY_BUILD_TYPES}` (and same for SDK overlay).  Having to expand this variable everywhere seems boilerplate-y.  Could we do something smarter, for example, infer the defaults if `IS_STDLIB` but none of the explicit keywords are specified, and if any keywords are specified, those would override the defaults?

**Comment by gribozavr** [comment]
@modocache OS X error:

```
CMake Error at cmake/modules/AddSwift.cmake:1121 (message):
  Either SHARED, STATIC, or OBJECT_LIBRARY must be specified
Call Stack (most recent call first):
  stdlib/public/SDK/Intents/CMakeLists.txt:1 (add_swift_library)
```

**Comment by modocache** [comment]
@swift-ci please test

**Comment by modocache** [comment]
@gribozavr Yes, that sounds like a good idea -- at some point I'll infer defaults based on `IS_STDLIB` and `IS_SDK_OVERLAY`.

I fixed the Intents SDK error -- my bad, I was working on an older branch, and on Linux. I also caught a new XPC SDK overlay was added (cool!), so I added the expansion to that call as well. Hopefully OS X tests will pass this time!

**Comment by modocache** [comment]
@swift-ci please test

**Comment by modocache** [comment]
@swift-ci please test

File: stdlib/public/SwiftRemoteMirror/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-add_swift_library(swiftRemoteMirror SHARED TARGET_LIBRARY DONT_EMBED_BITCODE
+add_swift_library(swiftRemoteMirror ${SWIFT_STDLIB_LIBRARY_BUILD_TYPES} TARGET_LIBRARY DONT_EMBED_BITCODE
```
Comment (modocache) [inline stdlib/public/SwiftRemoteMirror/CMakeLists.txt]: @gribozavr You mentioned we always want `libswiftRemoteMirror` to be built as a shared library. Does this mean this should be hardcoded as `SHARED`, and only built when `SWIFT_BUILD_STDLIB` is turned on?

Comment (gribozavr) [inline stdlib/public/SwiftRemoteMirror/CMakeLists.txt]: > Does this mean this should be hardcoded as `SHARED`, and only built when `SWIFT_BUILD_STDLIB` is turned on?

Yes to both, good catch!

**Comment by modocache** [comment]
@gribozavr Addressed the `libswiftRemoteMirror` issue. Thanks! Is this good to merge if the tests pass?

@swift-ci please test

**Comment by gribozavr** [comment]
@modocache Thank you, merged!

**Comment by modocache** [comment]
Awesome, thanks!!

**Pull request merged**